src/plugins/shared-variables/graph/plotted-variables-adornment/plotted-variables-adornment-component.tsx

```diff
@@ -215,50 +208,83 @@ export const PlottedVariablesAdornmentComponent = observer(function PlottedVaria
-  }, [dataConfig, model, plotWidth, plotHeight, sharedVariables, xAxis, yAxis, refreshValues]);
+    return mstReaction<string,boolean>(
+      () => {
+        if (xAxis && yAxis) {
+          const { domain: xDomain } = xAxis;
+          const { domain: yDomain } = yAxis;
+          // Return a primitive value (string, not array) so that equality will work
+          // and effect will only be called when these numbers actually change.
+          return `${xDomain[0]}, ${xDomain[1]}, ${yDomain[0]}, ${yDomain[1]}`;
+        }
+        return '';
+      },
+      (limits) => {
+        refreshValues();
+      },
+      { fireImmediately: true, name: "PlottedVariablesAdornmentComponent.refreshAxisChange" },
+      model);
+  }, [model, plotWidth, plotHeight, xAxis, yAxis, refreshValues, controller]);
+
+  // Refresh display when variable values change
+  useEffect(function udpateOnValuesChange() {
+    return mstReaction(
+      () => {
+        let hasAnyVars = false;
+        const varValues = Array.from(model.plottedVariables.values()).map(plottedVariables => {
+          if (plottedVariables.xVariable || plottedVariables.yVariable) {
+            hasAnyVars = true;
+          }
+          return {
+            x: plottedVariables.xVariable,
+            y: plottedVariables.yVariable?.computedValueIncludingMessageAndError
+          };
+        });
+        return {hasAnyVars, varValues};
+      },
+      (args: {hasAnyVars: boolean, varValues: any}) => {
+        if (args.hasAnyVars) {
+          if (settings.scalePlotOnValueChange && !graphModel.lockAxes) {
+            controller?.autoscaleAllAxes();
+          }
+          refreshValues();
+        }
+      },
+      { name: "PlottedVariablesAdornmentComponent.rescaleOnValuesChange" },
+      model);
+  // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [controller, model]);
 
   // Scale graph when a new X or Y variable is selected
   useEffect(function scaleOnVariableChange() {
     return mstReaction(() => {
       return Array.from(model.plottedVariables.values()).map((pvi) => [pvi.xVariableId, pvi.yVariableId]);
     },
       (varlist) => {
-        if (graphModel.lockAxes) return;
-        // Set a range that includes 0 to 2x for all the given values.
-        function fitValues(values: number[], axis: IAxisModel) {
-          if (values.length) {
-            setNiceDomain([0, ...values.map(x => 2 * x)], axis);
-          }
-        }
-
-        const variableValues = model.variableValues;
-        if (xAxis && yAxis) {
-          fitValues(variableValues.x, xAxis);
-          fitValues(variableValues.y, yAxis);
+        if (!graphModel.lockAxes) {
+          controller?.autoscaleAllAxes();
         }
       },
       { name: "PlottedVariablesAdornmentComponent.scaleOnVariableChange" },
       model);
-  }, [graphModel.lockAxes, model, xAxis, yAxis]);
+  }, [graphModel.lockAxes, model, xAxis, yAxis, controller]);
```

Comment:
In this case, I would use
```suggestion
  }, [graphModel, model, xAxis, yAxis, controller]);
```
A change in the value of `lockAxes` doesn't require reinstalling the reaction, but a change in the `graphModel` does.